File: knavwid/src/main/java/com/github/programmerr47/knavwid/AutoLayoutNavigationBuilder.java
package com.github.programmerr47.knavwid;

import android.view.View;
import com.github.programmerr47.knavwid.layoutfactory.DummyLayoutFactory;
import com.github.programmerr47.knavwid.layoutfactory.IdLayoutFactory;
import com.github.programmerr47.knavwid.layoutfactory.LayoutFactory;
import com.github.programmerr47.knavwid.layoutfactory.NavigationLayoutFactory;

public final class AutoLayoutNavigationBuilder extends NavigationBuilder<AutoLayoutNavigationBuilder> {
    private boolean includeToolbar;
    private boolean includeBottomBar;

    public static AutoLayoutNavigationBuilder navigation(int id) {
        return new AutoLayoutNavigationBuilder(new IdLayoutFactory(id));
    }

    public static AutoLayoutNavigationBuilder navigation(View view) {
        return new AutoLayoutNavigationBuilder(new DummyLayoutFactory(view));
    }

    public AutoLayoutNavigationBuilder(LayoutFactory layoutFactory) {
        super(layoutFactory, NavigationDefaults.NavigationDefaultsHolder.navigationDefaults());
    }

    @Override
    protected AutoLayoutNavigationBuilder getThis() {
        return this;
    }

    @Override
    public LayoutFactory layoutFactory() {
        return new NavigationLayoutFactory(includeToolbar, includeBottomBar, super.layoutFactory());
    }

    public AutoLayoutNavigationBuilder includeToolbar() {
        this.includeToolbar = true;
        return this;
    }

    public AutoLayoutNavigationBuilder excludeToolbar() {
        this.includeToolbar = false;
        return this;
    }

    public AutoLayoutNavigationBuilder includeBottomNavigation() {
        this.includeBottomBar = true;
        return this;
    }

    public AutoLayoutNavigationBuilder excludeBottomNavigation() {
        this.includeBottomBar = false;
        return this;
    }

    public CustomLayoutNavigationBuilder custom() {
        return new CustomLayoutNavigationBuilder(super.layoutFactory());
    }
}

File: knavwid/src/main/java/com/github/programmerr47/knavwid/AutoLayoutNavigationBuilder.kt
package com.github.programmerr47.knavwid

import android.view.View
import com.github.programmerr47.knavwid.layoutfactory.*

class AutoLayoutNavigationBuilder(layoutFactory: LayoutFactory) : NavigationBuilder<AutoLayoutNavigationBuilder>(
    layoutFactory,
    NavigationDefaults.NavigationDefaultsHolder.navigationDefaults()!!
) {
    private var includeToolbar: Boolean = false
    private var includeBottomBar: Boolean = false
    private var withoutShadow: Boolean = false

    private var tabTitleReses: IntArray? = null

    override val `this`: AutoLayoutNavigationBuilder get() = this

    override fun layoutFactory() = 
        NavigationLayoutFactory(includeToolbar, includeBottomBar, tabTitleReses, withoutShadow, super.layoutFactory())

    fun includeToolbar() = apply {
        this.includeToolbar = true
    }

    fun excludeToolbar() = apply {
        this.includeToolbar = false
    }

    fun withoutShadow() = apply {
        this.withoutShadow = true
    }

    fun tabs(vararg titleReses: Int) = apply {
        this.tabTitleReses = titleReses
    }

    fun includeBottomNavigation() = apply {
        this.includeBottomBar = true
    }

    fun excludeBottomNavigation() = apply {
        this.includeBottomBar = false
    }

    fun custom() = CustomLayoutNavigationBuilder(super.layoutFactory())

    companion object {

        fun navigation(id: Int) = AutoLayoutNavigationBuilder(IdLayoutFactory(id))
        fun navigation(view: View) = AutoLayoutNavigationBuilder(SimpleLayoutFactory(view))
    }
}
